### ISSUE TITLE
solutions for non-rectangular data

### ISSUE THREAD

**Issue by dylanbeaudette** [r-spatialecology/belg#4]
Thanks for putting this package together.

Do you have any suggestions on how to efficiently and consistently handle those cases where the raster data do not fit within a rectangular, NA-free region?

For example, I would like to use the `get_boltzmann` function on those pixels that are within management units, soil mapping units, and other such cases. Here is an example based on CTI values within a single delineation of a soil mapping unit:

![image](https://user-images.githubusercontent.com/624277/37309801-40ff3030-25ff-11e8-804e-13c7961b15fd.png)

The "white" space around the data are encoded as NA and thus  `get_boltzmann` throws an error. I can replace the NA values with some kind of indicator, but this seems like a hack, and affects the results.

Here is a short example.

```r
# no DATA, reference case
im <- matrix(sample(1:100, size=100, replace = TRUE), nrow = 10, ncol = 10)
r <- raster(im)
plot(r)
get_boltzmann(r)

# add NA, much more realistic scenario
im[which(upper.tri(im))] <- NA
r.2 <- raster(im)
plot(r.2)
# doesn't work
get_boltzmann(r.2)

# replace NA with some kind of NODATA flag
r.2[is.na(r.2)] <- -9999

# ack, the results depend on the value used to encode NA
get_boltzmann(r.2)
```

Any ideas?

**Comment by Nowosad** [comment]
Hi @dylanbeaudette. 
Thanks for the interest!
The original idea (and the paper) didn't consider data with NA and a simple solution with a constant value instead of NA is not going to work very well - the order of values matters in this method.
I have thought about a small modification to allow for NAs in rasters and will try to implement and test it (hopefully) at the end of this week.

**Comment by dylanbeaudette** [comment]
Howdy! It is hard not to be interested in such interesting ideas.

Thanks for the quick response. I would be happy to test and send examples over if that would be of any help.

**Comment by Nowosad** [comment]
Hi @dylanbeaudette, I've got good news - https://github.com/Nowosad/belg/pull/5.
NA values are now supported. It uses my own method in these cases - I'll try to document it in the next few days.
Can you try the function on some of your data?

``` r
library(raster)                                                            
#> Loading required package: sp
library(belg)                                                              
                                                                           
# no DATA, reference case                                                  
im <- matrix(sample(1:100, size=100, replace = TRUE), nrow = 10, ncol = 10)
r <- raster(im)                                                            
plot(r)                                                                    
```

![](https://i.imgur.com/cckdyK5.png)

``` r
get_boltzmann(r)                                                           
#> [1] 607.3673
                                                                           
# add NA, much more realistic scenario                                     
im[which(upper.tri(im))] <- NA                                             
r.2 <- raster(im)                                                          
plot(r.2)                                                                  
```

![](https://i.imgur.com/387nKee.png)

``` r
# also works!                                                              
get_boltzmann(r.2)                                                         
#> [1] 430.4155
```

**Comment by dylanbeaudette** [comment]
Great job! I'll need to move some data over to a computer with a real OS before I can test. This looks reasonable, however, it will be interesting to see how the results changes as the configuration of NA is adjusted. Is there any mechanism for distributing compiled binaries via GH?

**Comment by Nowosad** [comment]
This is a part of the main repo now, so you can just use:
```
# install.packages("devtools")
devtools::install_github("Nowosad/belg")
```

**Comment by dylanbeaudette** [comment]
Hmm, I get this on Windoz:

```
Error: running command '"C:/PROGRA~1/R/R-34~1.0/bin/x64/R" --no-site-file --no-environ --no-save --no-restore --quiet CMD config CC' had status 127
```

**Comment by Nowosad** [comment]
I've created a windows binary for "belg" - https://www.dropbox.com/s/fkiwmh54zbldwyl/belg_0.2.0.zip?dl=0
You should be able to install it on your computer (https://stackoverflow.com/questions/7075709/install-binary-zipped-r-package-via-command-line).

**Comment by dylanbeaudette** [comment]
Thanks!

I was able to compile from source on my Mac and examples seem to run as expected. I was able to run on some of my own data and the results _seem_ useful, but a rigorous evaluation will take some time.

I noticed an interesting relationship between the output from `get_boltzmann` and the area of the feature I was using to mask raster data. I'll post that tonight.

Since the functions now work with NA, OK to close issue?
